lib/public_key/src/pubkey_policy_tree.erl

```diff
@@ -0,0 +1,414 @@
+                end;
+           (#'PolicyQualifierInfo'{policyQualifierId = ?'id-qt-cps',
+                                   qualifier = Qualifier}) ->
+                {uri, public_key:der_decode('CPSuri', Qualifier)}
+        end,
+    Collect = fun(#{qualifier_set := QSet}) ->
+                      lists:map(FormatQualifier, QSet)
+              end,
+    case collect_subtree_qualifiers(Collect, SubTree, Policy) of
+        [] ->
+            collect_subtree_qualifiers(Collect, SubTree, ?anyPolicy);
+        QSet ->
+            QSet
+    end.
+
+%% --------------------------------------------------------------------
+-spec constraint_policy_node_set(policy_tree()) -> [public_key:policy_node()].
+
+%%
+%% From: PKITS.pdf:
+%% If the valid_policy_tree includes a leaf node with a valid_policy
+%% of anyPolicy, then the user-constrained-policy-set is
+%% any-policy. Otherwise, the user-constrained-policy-set consists of
+%% the set containing the valid_policy from each node in the
+%% valid_policy_tree in which the valid_policy is not anyPolicy and
+%% the valid_policy of that node's parent is anyPolicy.  The
+%% authorities-constrained-policy-set may be computed using the same
+%% procedure on the valid_policy_tree before its intersection with the
+%% user-initial-policy-set has calculated
+%%
+%% That is this function caclualates the valid policy node
+%% set after being constrained either by the user or only by the
+%% authorities (certificate extensions).
+%% --------------------------------------------------------------------
+constraint_policy_node_set({}) ->
+    [];
+constraint_policy_node_set(Tree) ->
+    case any_leaves(Tree) of
+        [] ->
+            constrain(Tree);
+        AnyLeaves ->
+            AnyLeaves
+    end.
+
+%%--------------------------------------------------------------------
+-spec empty() -> policy_tree().
+%%--------------------------------------------------------------------
+empty() ->
+    {}.
+
+%%--------------------------------------------------------------------
+-spec in_set(Policy::public_key:oid(), Set::[public_key:policy_node()]) -> boolean().
+%%
+%% Is there a node with <Policy> in <Set>
+%%--------------------------------------------------------------------
+in_set(_, []) ->
+    false;
+in_set(Policy, [#{valid_policy := Policy} |_]) ->
+    true;
+in_set(Policy, [_ | Rest]) ->
+    in_set(Policy, Rest).
+
+%%--------------------------------------------------------------------
+-spec is_empty(policy_tree()) -> boolean().
+%%--------------------------------------------------------------------
+is_empty({}) ->
+    true;
+is_empty(_) ->
+    false.
+
+%%--------------------------------------------------------------------
+-spec map_leaves(policy_tree(), LeafFun::function()) -> policy_tree().
+%%
+%% Update all leaves as determined by <LeafFun>
+%%--------------------------------------------------------------------
+map_leaves({Parent, [{_, _}|_] = SubTree}, LeafFun) ->
```

Comment:
in such places (function headers) I think it would be elegant and less error prone to have guard, like:
`map_leavs({Parent, [Child | _]}, LeafFun) when not is_leaf(Child) ->`

Reply:
We can not have user defined guards! I think it is ok that this module is implementation aware.

Reply:
you're right I guess.